includes/wikia/parser/templatetypes/RecognizedTemplatesProvider.class.php

```diff
@@ -0,0 +1,116 @@
+ * (both directly and indirectly)
+ * e.g. wgContentNamespaces
+ */
+
+class RecognizedTemplatesProvider {
+	const ERROR_MESSAGE = 'ExternalTemplatesProviderError';
+
+	private $tcs;
+	private $wikiaId;
+	private $namespaces;
+	private $namespacesTemplates;
+
+	public function __construct( TemplateClassificationService $tcs, $wikiaId, $namespaces = [] ) {
+		$this->tcs = $tcs;
+		$this->wikiaId = $wikiaId;
+		$this->namespaces = $namespaces;
+	}
+
+	public function getRecognizedTemplates() {
+		$recognizedTemplates = $this->getTemplates( true );
+		if ( empty( $this->namespaces ) ) {
+			return $recognizedTemplates;
+		}
+		return $this->intersectSets( $this->getNamespacesTemplates(), $recognizedTemplates );
+	}
+
+	public function getNotRecognizedTemplates() {
+		$notRecognizedTemplates = $this->getTemplates( false );
+		if ( empty( $this->namespaces ) ) {
+			return $notRecognizedTemplates;
+		}
+		return $this->intersectSets( $this->getNamespacesTemplates(), $notRecognizedTemplates );
+	}
+
+	public static function isRecognized( $type ) {
+		return $type !== TemplateClassificationService::TEMPLATE_UNKNOWN
+		&& $type !== AutomaticTemplateTypes::TEMPLATE_UNCLASSIFIED
+		&& $type !== AutomaticTemplateTypes::TEMPLATE_OTHER;
+	}
+
+	private function getTemplates( $getRecognized = true ) {
+		try {
+			$templates = $this->tcs->getTemplatesOnWiki( $this->wikiaId );
+		} catch ( \Swagger\Client\ApiException $exception ) {
+			$context = [ 'TCSApiException' => $exception ];
+			$this->handleException( $context );
+			return [];
+		}
+		foreach ( $templates as $pageId => $type ) {
+			$isRecognized = self::isRecognized( $type );
+			$shouldRemove = $getRecognized ? !$isRecognized : $isRecognized;
+			if ( $shouldRemove ) {
+				unset( $templates[$pageId] );
+			}
+		}
+		return $templates;
+	}
+
+	public function getNamespacesTemplates() {
+		if ( isset( $this->namespacesTemplates ) ) {
+			return $this->namespacesTemplates;
+		}
+
+		if ( empty( $this->namespaces ) ) {
+			return [];
+		}
+
+		$this->namespacesTemplates = ( new \WikiaSQL() )
+			->SELECT()->DISTINCT( 'p2.page_id as temp_id' )
+			->FROM( 'page' )->AS_( 'p' )
+			->INNER_JOIN( 'templatelinks' )->AS_( 't' )
+			->ON( 't.tl_from', 'p.page_id' )
+			->INNER_JOIN( 'page' )->AS_( 'p2' )
+			->ON( 'p2.page_title', 't.tl_title' )
+			->WHERE( 'p.page_namespace' )->IN( $this->namespaces )
+			->AND_( 'p2.page_namespace' )->EQUAL_TO( NS_TEMPLATE )
+			->AND_( 'p.page_id' )->NOT_EQUAL_TO( Title::newMainPage()->getArticleID() )
+			->runLoop( $this->getDB(), function ( &$pages, $row ) {
+				$pages[] = $row->temp_id;
+			} );
+
+		return $this->namespacesTemplates;
+	}
+
+	/**
+	 * Remove templates from second set that doesn't exist in first set
+	 */
+	private function intersectSets( $namespacesTemplates, $recognizedTemplates ) {
+		$namespacesRecognizedTemplates = [];
+		foreach ( $recognizedTemplates as $pageId => $type ) {
```

Comment:
I think that a far better solution to this problem is to construct the `$namespacesTemplates` array in a way that a page id is also a key, so e.g.:
``` [
    31 => 31,
    43 => 43
]```
and then using (array_intersect_key)[http://php.net/manual/en/function.array-intersect-key.php] to retrieve the common keys.